hermeto/core/package_managers/gomod.py

```diff
@@ -964,6 +967,83 @@ def _setup_go_toolchain(go_mod_file: RootedPath) -> Go:
     return go
 
 
+def _select_toolchain(go_mod_file: RootedPath, installed_toolchains: Iterable[Go]) -> Optional[Go]:
+    """
+    Pick the closest matching installed toolchain give a go.mod file.
+
+    :param go_mod_file: path to an application go.mod file as RootedPath
+    :param installed_toolchains: an iterable of Go instances pointing to actual Go binaries
+    :return: a Go instance which matches the go.mod version constraints or None if we could not find
+             a satisfying toolchain version installed
+    """
+    go_max_version = GoVersion.max()
+    go_mod_version_msg = "go.mod reported versions: '%s'[go], '%s'[toolchain]"
+
+    go_version_str, toolchain_version_str = _get_gomod_version(go_mod_file)
+    log.debug(
+        go_mod_version_msg,
+        go_version_str if go_version_str else "-",
+        toolchain_version_str if toolchain_version_str else "-",
+    )
+
+    if not go_version_str:
+        # Go added the 'go' directive to go.mod in 1.12 [1]. If missing, 1.16 is assumed [2].
+        # For our version comparison purposes we set the version explicitly to 1.20 if missing.
+        # [1] https://go.dev/doc/go1.12#modules
+        # [2] https://go.dev/ref/mod#go-mod-file-go
+        go_version_str = "1.20"
+        log.debug("Could not parse Go version from go.mod, using %s as fallback", go_version_str)
+
+    if not toolchain_version_str:
+        toolchain_version_str = go_version_str
+
+    go_mod_version = GoVersion(go_version_str)
+    go_mod_toolchain_version = GoVersion(toolchain_version_str)
+
+    if go_mod_version >= go_mod_toolchain_version:
+        target_version = go_mod_version
+    else:
+        target_version = go_mod_toolchain_version
+
+    if target_version.language > go_max_version.language:
+        raise PackageManagerError(
+            f"Required/recommended Go toolchain version '{target_version}' is not supported yet.",
+            solution=(
+                "Please lower your required/recommended Go version and retry the request. "
+                "You may also want to open a feature request on adding support for this version."
+            ),
+        )
+
+    # If we cannot find a matching toolchain, we'll try to fallback to a 1.21 one
+    matching_toolchains = filter(lambda t: t.version >= target_version, installed_toolchains)
+    try:
+        go = min(matching_toolchains)
+        log.debug("Using Go toolchain version '%s'", go.version)
+    except ValueError:
```

Comment:
Non-blocking nitpick: this one is a bit cryptic, maybe a comment like `# no toolchain satisfied the condition` could be added on this same line?